src/i965_encoder_vp8.c

```diff
@@ -0,0 +1,6244 @@
+            vp8_context->down_scaled_width_in_mb4x = vp8_context->min_scaled_dimension_in_mbs;
+        }
+
+        if (vp8_context->down_scaled_height_4x < vp8_context->min_scaled_dimension ||
+            vp8_context->down_scaled_height_in_mb4x < vp8_context->min_scaled_dimension_in_mbs) {
+
+            vp8_context->down_scaled_height_4x = vp8_context->min_scaled_dimension;
+            vp8_context->down_scaled_height_in_mb4x = vp8_context->min_scaled_dimension_in_mbs;
+        }
+    } else if (vp8_context->down_scaled_width_16x < vp8_context->min_scaled_dimension ||
+               vp8_context->down_scaled_width_in_mb16x < vp8_context->min_scaled_dimension_in_mbs ||
+               vp8_context->down_scaled_height_16x < vp8_context->min_scaled_dimension ||
+               vp8_context->down_scaled_height_in_mb16x < vp8_context->min_scaled_dimension_in_mbs) {
+
+        if (vp8_context->down_scaled_width_16x < vp8_context->min_scaled_dimension ||
+            vp8_context->down_scaled_width_in_mb16x < vp8_context->min_scaled_dimension_in_mbs) {
+
+            vp8_context->down_scaled_width_16x = vp8_context->min_scaled_dimension;
+            vp8_context->down_scaled_width_in_mb16x = vp8_context->min_scaled_dimension_in_mbs;
+        }
+
+        if (vp8_context->down_scaled_height_16x < vp8_context->min_scaled_dimension ||
+            vp8_context->down_scaled_height_in_mb16x < vp8_context->min_scaled_dimension_in_mbs) {
+
+            vp8_context->down_scaled_height_16x = vp8_context->min_scaled_dimension;
+            vp8_context->down_scaled_height_in_mb16x = vp8_context->min_scaled_dimension_in_mbs;
+        }
+    }
+}
+
+static void
+i965_encoder_vp8_free_surfaces(void **data)
+{
+    struct i965_encoder_vp8_surface *vp8_surface;
+
+    if (!data || *data)
+        return;
+
+    vp8_surface = *data;
+
+    if (vp8_surface->scaled_4x_surface_obj) {
+        i965_DestroySurfaces(vp8_surface->ctx, &vp8_surface->scaled_4x_surface_id, 1);
+        vp8_surface->scaled_4x_surface_id = VA_INVALID_SURFACE;
+        vp8_surface->scaled_4x_surface_obj = NULL;
+    }
+
+    if (vp8_surface->scaled_16x_surface_obj) {
+        i965_DestroySurfaces(vp8_surface->ctx, &vp8_surface->scaled_16x_surface_id, 1);
+        vp8_surface->scaled_16x_surface_id = VA_INVALID_SURFACE;
+        vp8_surface->scaled_16x_surface_obj = NULL;
+    }
+}
+
+static void
+i965_encoder_vp8_allocate_surfaces(VADriverContextP ctx,
+                                   struct intel_encoder_context *encoder_context,
+                                   struct object_surface *obj_surface)
+{
+    struct i965_driver_data *i965 = i965_driver_data(ctx);
+    struct i965_encoder_vp8_context *vp8_context = encoder_context->vme_context;
+    struct i965_encoder_vp8_surface *vp8_surface;
+    int down_scaled_width_4x, down_scaled_height_4x;
+    int down_scaled_width_16x, down_scaled_height_16x;
+
+    if (!obj_surface || obj_surface->private_data)
```

Comment:
I double checked the issue when sharing a surface between two different contexts with yamitrancode on SKL
1.  converting vp8 to avc/vp8/hevc works fine
     The output of VP8 encoder doesn't have private surface, so converting vp8 to any other codec always works.  

2.  converting hevc/vc1/avc to vp8/avc works fine
     VP8 and AVC encoder doesn't read/write any data from/to the private data structure of the output of HEVC/VC1/AVC decoder, so it also works.

3.  converting vc1/avc to hevc results segmentation fault
HEVC encoder read/writes some data from/to the private data structure of the ouput private data structure of  VC1/AVC decoder, which results in the segmentation fault issue.

So what we need to do is to fix HEVC encoder.

In addition, I think we should not share surfaces between different contexts except the input/output surface, otherwise it will make the application and driver more complicated.